demos/link-prediction/random-walks/ctdne-link-prediction.ipynb

```diff
@@ -0,0 +1,680 @@
+      "text/plain": [
+       "<Figure size 504x504 with 1 Axes>"
+      ]
+     },
+     "metadata": {
+      "needs_background": "light"
+     },
+     "output_type": "display_data"
+    }
+   ],
+   "source": [
+    "temporal_node_embeddings = temporal_model.wv.vectors\n",
+    "static_node_embeddings = static_model.wv.vectors\n",
+    "plot_tsne(\"TSNE visualisation of temporal node embeddings\", temporal_node_embeddings)\n",
+    "plot_tsne(\"TSNE visualisation of static node embeddings\", static_node_embeddings)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Link Prediction using Node Embeddings\n",
+    "\n",
+    "The node embeddings we've generated can now be used as input for a link prediction task. The reported AUC scores from the reference paper for this dataset are:\n",
+    "\n",
+    "* Node2Vec: 0.759 \n",
+    "* CTDNE: 0.777\n",
+    "\n",
+    "Below are a set of helper functions we can use to train and evaluate a link prediction classifier. The rest of the notebook will use the binary operator defined in the cell below. Available operators in the demo are:\n",
+    "\n",
+    "* `hadamard`\n",
+    "* `avg`\n",
+    "* `l1`\n",
+    "* `l2` \n",
+    "\n",
+    "Alternatively, other user defined function taking two node embeddings to produce a link embedding could be used, but may affect convergence of the classifier model."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 13,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "binary_operator = \"l2\""
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 14,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "from sklearn.pipeline import Pipeline\n",
+    "from sklearn.linear_model import LogisticRegressionCV\n",
+    "from sklearn.metrics import roc_auc_score\n",
+    "from sklearn.preprocessing import StandardScaler\n",
+    "\n",
+    "\n",
+    "def operator_hadamard(u, v):\n",
+    "    return u * v\n",
+    "\n",
+    "\n",
+    "def operator_avg(u, v):\n",
+    "    return (u + v) / 2.0\n",
+    "\n",
+    "\n",
+    "def operator_l1(u, v):\n",
+    "    return np.abs(u - v)\n",
+    "\n",
+    "\n",
+    "def operator_l2(u, v):\n",
+    "    return (u - v) ** 2\n",
+    "\n",
+    "\n",
+    "operator_func = {\n",
+    "    \"hadamard\": operator_hadamard,\n",
+    "    \"avg\": operator_avg,\n",
+    "    \"l1\": operator_l1,\n",
+    "    \"l2\": operator_l2,\n",
+    "}\n",
+    "\n",
+    "\n",
+    "def link_examples_to_features(link_examples, transform_node, binary_operator):\n",
+    "    op_func = (\n",
+    "        operator_func[binary_operator]\n",
+    "        if isinstance(binary_operator, str)\n",
+    "        else binary_operator\n",
+    "    )\n",
+    "    return [\n",
+    "        op_func(transform_node(src), transform_node(dst)) for src, dst in link_examples\n",
+    "    ]\n",
+    "\n",
+    "\n",
+    "def link_prediction_clf(max_iter=2000):\n",
```

Comment:
What's CLF stand for? I see it's used in the sklearn docs too, but... it's not obvious to me.

Reply:
I assumed "classifier"?